File: modules/raft/src/test/java/org/apache/ignite/raft/jraft/util/UtilsTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ignite.raft.jraft.util;

import java.nio.ByteBuffer;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import org.apache.ignite.raft.jraft.Closure;
import org.apache.ignite.raft.jraft.Status;
import org.apache.ignite.raft.jraft.error.RaftError;
import org.junit.Assert;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;

/**
 *
 */
public class UtilsTest {
    private Executor executor = Executors.newSingleThreadExecutor();

    @Test
    public void testRunThread() throws Exception {
        CountDownLatch latch = new CountDownLatch(1);

        Utils.runInThread(executor, new Runnable() {
            @Override
            public void run() {
                latch.countDown();
            }
        });
        latch.await();
    }

    @Test(expected = IllegalArgumentException.class)
    public void tetsVerifyGroupId1() {
        Utils.verifyGroupId("");
    }

    @Test(expected = IllegalArgumentException.class)
    public void tetsVerifyGroupId2() {
        Utils.verifyGroupId(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void tetsVerifyGroupId3() {
        Utils.verifyGroupId("1abc");
    }

    @Test(expected = IllegalArgumentException.class)
    public void tetsVerifyGroupId4() {
        Utils.verifyGroupId("*test");
    }

    @Test
    public void tetsVerifyGroupId5() {
        Utils.verifyGroupId("t");
        Utils.verifyGroupId("T");
        Utils.verifyGroupId("Test");
        Utils.verifyGroupId("test");
        Utils.verifyGroupId("test-hello");
        Utils.verifyGroupId("test123");
        Utils.verifyGroupId("t_hello");
    }

    @Test
    public void testRunClosure() throws Exception {
        CountDownLatch latch = new CountDownLatch(1);
        Utils.runClosureInExecutor(executor, new Closure() {

            @Override
            public void run(Status status) {
                assertTrue(status.isOk());
                latch.countDown();
            }
        });
        latch.await();
    }

    @Test
    public void testRunClosureWithStatus() throws Exception {
        CountDownLatch latch = new CountDownLatch(1);
        Utils.runClosureInExecutor(executor, new Closure() {

            @Override
            public void run(Status status) {
                assertFalse(status.isOk());
                Assert.assertEquals(RaftError.EACCES.getNumber(), status.getCode());
                assertEquals("test 99", status.getErrorMsg());
                latch.countDown();
            }
        }, new Status(RaftError.EACCES, "test %d", 99));
        latch.await();
    }

    @Test
    public void test_getProcessId() {
        long pid = Utils.getProcessId(-1);
        assertNotEquals(-1, pid);
        System.out.println("test pid:" + pid);
    }

    @Test
    public void testAllocateExpandByteBuffer() {
        ByteBuffer buf = Utils.allocate(128);
        assertEquals(0, buf.position());
        assertEquals(128, buf.capacity());
        assertEquals(128, buf.remaining());

        buf.put("hello".getBytes());
        assertEquals(5, buf.position());

        buf = Utils.expandByteBufferAtLeast(buf, 128);
        assertEquals(5, buf.position());
        assertEquals(1152, buf.capacity());
        assertEquals(1147, buf.remaining());

        buf = Utils.expandByteBufferAtLeast(buf, 2048);
        assertEquals(5, buf.position());
        assertEquals(1152 + 2048, buf.capacity());
        assertEquals(1147 + 2048, buf.remaining());
    }

    @Test
    public void testParsePeerId() {
        String pid = "192.168.1.88:5566";
        String[] result = Utils.parsePeerId(pid);
        String[] expecteds = {"192.168.1.88", "5566"};
        Assert.assertTrue(result.length == 2);
        Assert.assertArrayEquals(expecteds, result);

        pid = "[fe80:0:0:0:6450:aa3c:cd98:ed0f]:8847";
        result = Utils.parsePeerId(pid);
        expecteds = new String[] {"[fe80:0:0:0:6450:aa3c:cd98:ed0f]", "8847"};
        Assert.assertTrue(result.length == 2);
        Assert.assertArrayEquals(expecteds, result);

        pid = "192.168.1.88:5566:9";
        result = Utils.parsePeerId(pid);
        expecteds = new String[] {"192.168.1.88", "5566", "9"};
        Assert.assertTrue(result.length == 3);
        Assert.assertArrayEquals(expecteds, result);

        pid = "[fe80:0:0:0:6450:aa3c:cd98:ed0f]:8847:9";
        result = Utils.parsePeerId(pid);
        expecteds = new String[] {"[fe80:0:0:0:6450:aa3c:cd98:ed0f]", "8847", "9"};
        Assert.assertTrue(result.length == 3);
        Assert.assertArrayEquals(expecteds, result);

        pid = "192.168.1.88:5566:0:6";
        result = Utils.parsePeerId(pid);
        expecteds = new String[] {"192.168.1.88", "5566", "0", "6"};
        Assert.assertTrue(result.length == 4);
        Assert.assertArrayEquals(expecteds, result);

        pid = "[fe80:0:0:0:6450:aa3c:cd98:ed0f]:8847:0:6";
        result = Utils.parsePeerId(pid);
        expecteds = new String[] {"[fe80:0:0:0:6450:aa3c:cd98:ed0f]", "8847", "0", "6"};
        Assert.assertTrue(result.length == 4);
        Assert.assertArrayEquals(expecteds, result);

        boolean ex1 = false;
        try {
            pid = "[192.168.1].88:eee:x:b:j";
            Utils.parsePeerId(pid);
        }
        catch (Exception e) {
            ex1 = true;
        }
        Assert.assertTrue(ex1);

        boolean ex2 = false;
        try {
            pid = "[dsfsadf]:eee:x:b:j";
            Utils.parsePeerId(pid);
        }
        catch (Exception e) {
            ex2 = true;
        }
        Assert.assertTrue(ex2);

    }

}

File: modules/platforms/cpp/ignite/client/detail/utils_test.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ignite/client/detail/utils.h"

#include <gtest/gtest.h>

using namespace ignite;
using namespace detail;

inline column make_column(std::string name, ignite_type type, std::int32_t key_index = -1) {
    column val;
    val.name = std::move(name);
    val.type = type;
    val.key_index = key_index;

    return val;
}

std::shared_ptr<schema> make_test_schema() {
    std::vector<column> columns;
    columns.push_back(make_column("VAL_COL1", ignite_type::INT32));
    columns.push_back(make_column("KEY_COL1", ignite_type::STRING, 0));
    columns.push_back(make_column("VAL_COL2", ignite_type::STRING));
    columns.push_back(make_column("KEY_COL2", ignite_type::INT32, 1));

    return schema::create_instance(0, std::move(columns));
}

ignite_tuple write_read_tuple(const ignite_tuple &tuple, const std::shared_ptr<schema> &sch, bool key_only) {
    std::vector<std::byte> message;
    protocol::buffer_adapter buffer(message);

    protocol::writer writer(buffer);

    write_tuple(writer, *sch, tuple, key_only);

    protocol::reader reader(message);
    reader.skip(); // Skip bitset

    return read_tuple(reader, sch.get(), key_only);
}

TEST(client_utils, tuple_write_read_random_order_all_columns) {
    auto sch = make_test_schema();

    ignite_tuple tuple{{"VAL_COL1", std::int32_t(42)}, {"VAL_COL2", std::string("Lorem ipsum")},
        {"KEY_COL2", std::int32_t(1337)}, {"KEY_COL1", std::string("Test value")}};

    auto res_tuple = write_read_tuple(tuple, sch, false);

    ASSERT_EQ(4, res_tuple.column_count());
    EXPECT_EQ("VAL_COL1", res_tuple.column_name(0));
    EXPECT_EQ("KEY_COL1", res_tuple.column_name(1));
    EXPECT_EQ("VAL_COL2", res_tuple.column_name(2));
    EXPECT_EQ("KEY_COL2", res_tuple.column_name(3));

    EXPECT_EQ(std::int32_t(42), res_tuple.get(0));
    EXPECT_EQ(std::string("Test value"), res_tuple.get(1));
    EXPECT_EQ(std::string("Lorem ipsum"), res_tuple.get(2));
    EXPECT_EQ(std::int32_t(1337), res_tuple.get(3));
}

TEST(client_utils, tuple_write_read_random_order_key_only) {
    auto sch = make_test_schema();

    ignite_tuple tuple{{"VAL_COL1", std::int32_t(42)}, {"VAL_COL2", std::string("Lorem ipsum")},
        {"KEY_COL2", std::int32_t(1337)}, {"KEY_COL1", std::string("Test value")}};

    auto res_tuple = write_read_tuple(tuple, sch, true);

    ASSERT_EQ(2, res_tuple.column_count());
    EXPECT_EQ("KEY_COL1", res_tuple.column_name(0));
    EXPECT_EQ("KEY_COL2", res_tuple.column_name(1));

    EXPECT_EQ(std::string("Test value"), res_tuple.get(0));
    EXPECT_EQ(std::int32_t(1337), res_tuple.get(1));
}
